libs/ui/src/session_time_limit_tracker.tsx

```diff
@@ -0,0 +1,206 @@
+/* istanbul ignore file */
+import pluralize from 'pluralize';
+import React, { useState } from 'react';
+import { useIdleTimer } from 'react-idle-timer';
+import styled from 'styled-components';
+import {
+  DEFAULT_INACTIVE_SESSION_TIME_LIMIT_MINUTES,
+  DEFAULT_OVERALL_SESSION_TIME_LIMIT_HOURS,
+  DippedSmartCardAuth,
+  InsertedSmartCardAuth,
+  UnixTimestampInMilliseconds,
+} from '@votingworks/types';
+
+import { Button } from './button';
+import { useNow } from './hooks/use_now';
+import { Modal } from './modal';
+import { Prose } from './prose';
+import { Timer } from './timer';
+
+const SECONDS_TO_WRAP_UP_AFTER_INACTIVE_SESSION_TIME_LIMIT = 60;
+const SECONDS_AHEAD_OF_OVERALL_SESSION_TIME_LIMIT_TO_WARN = 15 * 60;
+
+type AuthStatusWithSessionExpiry =
+  | DippedSmartCardAuth.RemoveCard
+  | DippedSmartCardAuth.LoggedIn
+  | InsertedSmartCardAuth.LoggedIn;
+
+function computeSecondsToSessionExpiry(
+  authStatus: AuthStatusWithSessionExpiry,
+  now: Date
+): number {
+  return Math.max(
+    (new Date(authStatus.sessionExpiresAt).getTime() - now.getTime()) / 1000,
+    0
+  );
+}
+
+function shouldDisplayTimeLimitPrompt(
+  authStatus: AuthStatusWithSessionExpiry,
+  now: Date
+): boolean {
+  const secondsToSessionExpiry = computeSecondsToSessionExpiry(authStatus, now);
+  return (
+    // Despite looking like a condition for only the overall session time limit, this condition
+    // also covers the inactive session time limit, since we trigger the logout for that limit by
+    // bringing the session expiry up, even closer than
+    // SECONDS_AHEAD_OF_OVERALL_SESSION_TIME_LIMIT_TO_WARN
+    secondsToSessionExpiry <=
+    SECONDS_AHEAD_OF_OVERALL_SESSION_TIME_LIMIT_TO_WARN
+  );
+}
+
+interface SessionTimeLimitTrackerHelperProps {
+  authStatus: AuthStatusWithSessionExpiry;
+  logOut: () => void;
+  updateSessionExpiry: (sessionExpiresAt: UnixTimestampInMilliseconds) => void;
+}
+
+/**
+ * A helper for SessionTimeLimitTracker
+ */
+function SessionTimeLimitTrackerHelper({
+  authStatus,
+  logOut,
+  updateSessionExpiry,
+}: SessionTimeLimitTrackerHelperProps): JSX.Element | null {
+  // TODO: Pull values from system settings
+  const inactiveSessionTimeLimitMinutes =
+    DEFAULT_INACTIVE_SESSION_TIME_LIMIT_MINUTES;
+  const overallSessionTimeLimitHours = DEFAULT_OVERALL_SESSION_TIME_LIMIT_HOURS;
+
+  const now = useNow().toJSDate();
```

Comment:
Yes good question! I'm using `useNow` here because it returns an updated date every second, triggering a re-render every second and making sure we constantly check whether it's time to surface the prompt or not:

https://github.com/votingworks/vxsuite/blob/04631463a3c68e8796a24f4c1bda114e234042e2/libs/ui/src/hooks/use_now.ts#L11-L13

With a static `new Date()`, we'd potentially never surface the prompt (unless a re-render was triggered by some other external factor). I'd have to instead set up a `setInterval` or `setTimeout` myself to make that happen. With `useNow`, I can get that for free.

In the `useIdleTimer` `onIdle` function, I don't need something like `useNow` because it's an event handler. So whenever it's called, `new Date()` will be up to date.

Reply:
Thanks!